### PR TITLE
Add trigger command for burst mode

### PR DESCRIPTION
**Pull request by couger01**

### CODE REVIEW

File: python/lsst/ts/tunablelaser/component.py [hunk 1]
```diff
@@ -121,9 +129,20 @@ async def set_output_energy_level(self, output_energy_level):
         self.log.debug(f"Changing output energy level={output_energy_level}")
         await self.m_cpu800.set_output_energy_level(output_energy_level)
 
+    async def set_trigger_mode(self):
+        """Set trigger mode for laser."""
```
Comment (r-owen) [inline python/lsst/ts/tunablelaser/component.py:133]: Please clarify the difference between this and set_burst_mode

Also please use a separate detailed state for these two modes (if they really are importantly different from a SAL user's perspective)

Comment (r-owen) [reply python/lsst/ts/tunablelaser/component.py:133]: What does this method do if the laser is not in burst mode? Does the user get an informative error? If not,  please fix that (if necessary, by checking the mode yourself before commanding the laser).

**Review by r-owen — CHANGES_REQUESTED** [review]
I may misunderstand the sequencing for bursts, but I think it is:
* put the laser into burst mode
* trigger as many bursts as desired
* take the laser out of burst mode

If so, I hope you will consider simplifying the API to reflect that.

File: python/lsst/ts/tunablelaser/component.py [hunk 2]
```diff
@@ -138,12 +157,30 @@ async def set_burst_count(self, count):
             The amount to pulse the laser.
         """
         await self.m_cpu800.set_burst_count(count)
+        await self.csc.evt_burstCountSet.set_write(count=count)
 
     async def start_propagating(self):
         """Start propagating the beam of the laser."""
         await self.m_cpu800.start_propagating()
+        await asyncio.sleep(10)  # laser warmup delay
```
Comment (r-owen) [inline python/lsst/ts/tunablelaser/component.py:165]: I strongly suggest making this delay a named constant (or a CSC configuration parameter if we are likely to need to tune it).

File: python/lsst/ts/tunablelaser/component.py [hunk 3]
```diff
@@ -138,12 +157,30 @@ async def set_burst_count(self, count):
...
+            == Mode.BURST
+        ):
+            await self.csc.publish_new_detailed_state(
+                LaserDetailedState.PROPAGATINGBURSTMODEWAITINGFORTRIGGER
+            )
+        elif (
+            self.m_cpu800.continous_burst_mode_trigger_burst_register.register_value
+            == Mode.CONTINUOUS
+        ):
+            await self.csc.publish_new_detailed_state(LaserDetailedState.PROPAGATING)
```
Comment (r-owen) [inline python/lsst/ts/tunablelaser/component.py:177]: Please handle the case that `self.m_cpu800.continous_burst_mode_trigger_burst_register.register_value` is neither of these two values. Right now you silently ignore it by not publishing a new detailed state and still setting self.is_propagating=True.

Is it possible to check this register *before* calling start_propagating? If so, I suggest doing that -- only turn on the laser if the mode is recognizable!

I gently suggest reducing duplication by setting a variable to the desired detailed state. For example (if/else or a match statement would also work fine, but need an else/_ clause to deal with an unexpected register value -- and either could produce a clearer error message than the KeyError that the following code would produce):
```
detailed_state = {
    Mode.CONTINUOUS: LaserDetailedState.PROPAGATING,
    Mode.BURST: LaserDetailedState.PROPAGATINGBURSTMODEWAITINGFORTRIGGER,
}[self.m_cpu800.continous_burst_mode_trigger_burst_register.register_value]
await self.csc.publish_new_detailed_state(detailed_state)
self.is_propagating = True
```

File: python/lsst/ts/tunablelaser/component.py [hunk 4]
```diff
@@ -138,12 +157,30 @@ async def set_burst_count(self, count):
...
         self.is_propgating = True
 
+    async def start_trigger(self):
+        """Trigger the laser's burst."""
+        await self.set_trigger_mode()
```
Comment (r-owen) [inline python/lsst/ts/tunablelaser/component.py:182]: Having two methods that do the same thing is confusing. I suggest you delete this and look at the comments on set_trigger_mode above.

File: python/lsst/ts/tunablelaser/component.py (hunk 1 above)
Comment (r-owen) [reply python/lsst/ts/tunablelaser/component.py:133]: These comments still apply and please delete the `start_trigger` method below.

File: python/lsst/ts/tunablelaser/component.py [hunk 5]
```diff
@@ -121,9 +129,20 @@ async def set_output_energy_level(self, output_energy_level):
...
+        """Set trigger mode for laser."""
+        await self.m_cpu800.set_propagation_mode(Mode.TRIGGER)
+        await self.csc.publish_new_detailed_state(
+            LaserDetailedState.PROPAGATINGBURSTMODETRIGGERED
+        )
+        await self.csc.publish_new_detailed_state(
+            LaserDetailedState.PROPAGATINGBURSTMODEWAITINGFORTRIGGER
+        )
+
+    async def set_burst_mode(self, count):
```
Comment (r-owen) [inline python/lsst/ts/tunablelaser/component.py:142]: Thank you for consolidating this. At risk of making you wish you had not...

Please document the `count` parameter. What is the valid range?

What does this do if count is out of range (< 1 or too big). Does self.m_cpu800.set_burst_count catch this and raise a clear error? If not, please check it here and raise an error before calling self.m_cpu800.set_burst_count

Please document the exception in an Exceptions section. (I suppose any command that calls self.m_cpu800.x may raise if the laser is not connected or something. It seems excessive to document those exceptions here, but the exception for count out of range is less obvious.)

**Review by r-owen — CHANGES_REQUESTED** [review]
Looks a lot better. I still hope you can get the # of "trigger a burst" methods down from 2 to 1 and name it something like "trigger_burst" or at least a name that does not include "set" or "mode".

File: python/lsst/ts/tunablelaser/component.py (hunk 3 above)
Comment (couger01) [reply python/lsst/ts/tunablelaser/component.py:177]: So in the AsciiRegister class, the set_register_value checks for correct values and raises a ValueError exception if it is not correct. The modes are only set by the 2 set_*_mode methods and therefore are validated by enum and then by AsciiRegister set_value method. I have added documentation to clarify that this exception will be raised.

Comment (r-owen) [reply python/lsst/ts/tunablelaser/component.py:206]: That does not handle the case that the register has been set by other code or corrupted by an error in the laser. I still think you should handle invalid values -- preferably before turning on the laser!

File: python/lsst/ts/tunablelaser/hardware.py [hunk 6]
```diff
@@ -300,7 +300,12 @@ async def start_propagating(self):
         None
 
         """
-        await self.power_register_2.set_register_value(Power.ON)
+        await self.continous_burst_mode_trigger_burst_register.read_register_value()
+        rv = self.continous_burst_mode_trigger_burst_register.register_value
+        if rv in list(Mode):
+            await self.power_register_2.set_register_value(Power.ON)
+        else:
+            raise RuntimeError(f"{rv} is not in one of accepted values")
```
Comment (r-owen) [inline python/lsst/ts/tunablelaser/hardware.py:308]: Nice. This seems like a really good place to put the test.

**Review by r-owen — APPROVED** [review]
Looks really good. I had a few minor suggestions I hope you will consider before merging.

**Pull request merged**